framework/Models/ExternalModel.py

```diff
@@ -222,54 +222,63 @@ def _externalRun(self, Input):
-    #else:
-    #  InputDict = Input
+      # preference is to make it a member of the insantiated object
+      if key.isidentifier():
+        setattr(externalSelf, key, copy.copy(modelVariableValues[key]))
+      # otherwise, just have it available in the dictionary instead. This feels redundant.
+      else:
+        self.raiseAWarning(f'Variable "{key}" could not be added to ExternalModel first argument due to complex name. ' +
+                            'Find it in the second (dictionary) argument instead.')
     # only pass the variables and their values according to the model itself.
-    for key in Input.keys():
-      if key in self.modelVariableType.keys() or key in additionalKeys:
+    for key in Input:
+      if key in self.modelVariableType or key in additionalKeys:
         InputDict[key] = Input[key]
 
+    # run the user's model
     self.sim.run(externalSelf, InputDict)
 
+    # collect the user results
     for key in self.modelVariableType:
-      try:
-        # Note, the following string can't be converted using {} formatting, at least as far as I can tell.
-        CustomCommandExecuter.execCommand('object["'+key+'"]  = copy.copy(self.'+key+')', self=externalSelf,object=modelVariableValues) #exec('modelVariableValues[key]  = copy.copy(externalSelf.'+key+')') #self.__pointSolution()
-      except (SyntaxError, AttributeError):
-        self.raiseAWarning('Variable "{}" cannot be read from "self" due to complex name.  Retaining original value.'.format(key))
-    for key in self.initExtSelf.__dict__.keys():
-      # Note, the following string can't be converted using {} formatting, at least as far as I can tell.
-      CustomCommandExecuter.execCommand('self.' +key+' = copy.copy(object.'+key+')', self=self.initExtSelf, object=externalSelf) #exec('self.initExtSelf.' +key+' = copy.copy(externalSelf.'+key+')')
+      # preferentially obtain values from the instantiated object
+      if key.isidentifier(): # checks valid python variable name
+        modelVariableValues[key] = copy.copy(getattr(externalSelf, key))
+      # otherwise, try to get it from the dictionary. If not, save it as "None" so it reports back to the user.
+      else:
+        self.raiseAWarning(f'Variable "{key}" could not be read from ExternalModel first argument due to complex name. ' +
+                            'Checking it in the second (dictionary) argument instead.')
+        modelVariableValues[key] = copy.copy(InputDict.get(key, None))
+    # in case initialization variables have been modified, update them.
+    for key in self.initExtSelf.__dict__:
+      if key.isidentifier(): # checks valid python variable name
+        setattr(self.initExtSelf, key, copy.copy(getattr(externalSelf, key)))
+      else:
+        self.raiseAnError(RuntimeError, f'Variable "{key}" has an invalid Python variable name and cannot be set in ExternalModel!')
```

Comment:
Do we need the if check here, we are looping through the dictionary of the object "self.initExtSelf"? Does key always satisfy  isidentifier here?

Reply:
You're probably right, there's probably not a chance of this being invalid. I can remove the conditional.